torch/tensortypes/nestedtensor/nested.py

```diff
@@ -0,0 +1,215 @@
+            if not torch.is_tensor(data_):
+                raise ValueError("Each element of the tuple or list must "
+                                 "be a torch.Tensor")
+        tensors = []
+        for data_ in data:
+            # torch.tensor copies on construction
+            new_data = torch.empty_like(data_)
+            new_data.copy_(data_)
+            new_data = new_data.to(dtype)
+            new_data = new_data.to(device)
+            new_data = new_data.requires_grad_(requires_grad)
+            if pin_memory:
+                new_data = new_data.pin_memory()
+            tensors.append(new_data)
+
+        return NestedTensor(tensors)
+
+def as_nestedtensor(data, dtype=None, device=None):
+    ret = NestedTensor(data)
+    if dtype is not None:
+        ret = ret.to(dtype)
+    if device is not None:
+        ret = ret.to(device)
+    return ret
+
+class NestedTensor():
+    # The attributes must match across all constiuents
+    # and default to the empty Tensor's if the given list
+    # is empty.
+    #
+    # The NestedTensor's attributes then become that of its
+    # constiuents.
+    #
+    # Attributes:
+    #     dim
+    #     layout
+    #     device
+    #     dtype
+    #     requires_grad
+    #     is_pinned
+    def __init__(self, tensors):
+        for tensor in tensors:
+            assert torch.is_tensor(tensor)
+        self.tensors = tensors
+        if len(tensors):
+            self.dim = tensors[0].dim()
+            self.layout = tensors[0].layout
+            self.device = tensors[0].device
+            self.dtype = tensors[0].dtype
+            requires_grad = tensors[0].requires_grad
+            is_pinned = tensors[0].is_pinned()
+            for tensor in tensors:
+                if not (self.dim == tensor.dim() and
+                        self.layout == tensor.layout and
+                        self.device == tensor.device and
+                        self.dtype == tensor.dtype and
+                        requires_grad == tensor.requires_grad and
+                        is_pinned == tensor.is_pinned()):
+                    raise ValueError("Each passed Tensor "
+                                     "must match in dim, layout, "
+                                     "device, dtype and requires_grad")
+        else:
+            # Carrying around information as member variables vs.
+            # checking one entry of the owned Tensors is annoying
+            # and error-prone. Carrying around an is_empty attribute
+            # to hide the fact that we carry around a list with a
+            # single empty Tensor is also annoying and error-prone.
+            # Both are not worth it for a minor feature.
+            raise ValueError("We do not support empty lists for now.")
+
+    def __getattribute__(self, attr):
+        if attr == 'shape':
+            raise NotImplementedError()
+        if attr == 'requires_grad':
+            return self.tensors[0].requires_grad
```

Comment:
This looks questionable. Wouldn't any(t.requires_grad for t in self.tensors) be a more plausible designation here?

Reply:
The invariant is that if one tensor requires_grad then all of the tensors inside the nested_tensor must require_grad

Reply:
I see that in the constructor you assert that the requires grad of all tensors must be the same on construction. But I can mutate the requires grad of a tensor in place, so....

Reply:
Other silly ways we can violate internal invariants of NestedTensor include dim-changing inplace resize on constituent tensor.

Reply:
> I see that in the constructor you assert that the requires grad of all tensors must be the same on construction. But I can mutate the requires grad of a tensor in place, so....

I think this depends on if if there is a way for the user to get at the constituent tensors directly or not.

Reply:
You're not supposed to use self.tensors - I'll change it to self.tensors_. But it's a private variable.

Reply:
> You're not supposed to use self.tensors - I'll change it to self.tensors_. But it's a private variable.

Python convention is that leading underscore is private